### DISCUSSION TITLE
How to read an irregular csv file?

### DISCUSSION THREAD

**Discussion by yiyinghu97** [duckdb/duckdb#12595] [Q&A]
How can I read a csv file like this. 
c1,c2
1,2,
3,4,
The expected result is :
c1 c2 column2
1   2
3   4

**Comment by szarnyasg** [comment]
Hi @yiyinghu97,

You can e.g. use:
```sql
select column0 as c1, column1 as c2, column2 as column2
from 'my.csv';
```

This returns:
```text
┌───────┬───────┬─────────┐
│  c1   │  c2   │ column2 │
│ int64 │ int64 │ varchar │
├───────┼───────┼─────────┤
│     1 │     2 │ NULL    │
│     3 │     4 │ NULL    │
└───────┴───────┴─────────┘
```

I cannot deduce all your requirements from the example though – would you like to see `NULL` values in `column2`? Is there a single column missing in the header? In many cases, it may be easier to not use the header, skip it, and manually specify the schema:

```sql
from read_csv('my.csv', header=false, skip=1, delim=',', columns={'c1': 'INTEGER', 'c2': 'INTEGER', 'column2': 'VARCHAR'});
```
```
┌───────┬───────┬─────────┐
│  c1   │  c2   │ column2 │
│ int32 │ int32 │ varchar │
├───────┼───────┼─────────┤
│     1 │     2 │ NULL    │
│     3 │     4 │ NULL    │
└───────┴───────┴─────────┘
```

**Reply by yiyinghu97** [reply]
Hi,  @szarnyasg 
Thanks for reply! But actually I don't know the schema of the file. 
I'm developing a file import system. Sometimes every line in the file except the first line has an extra comma, other times they don't. I want to be compatible with both situations.

**Comment by szarnyasg** [comment]
To handle the extra comma, check out the [`null_padding` attribute](https://duckdb.org/docs/data/csv/overview#parameters).

**Reply by yiyinghu97** [reply]
Thanks  so much! That works. But I probably find a bug. When I set "parallel = false". There is an extra line with all null value.
```
D Select * From read_csv('D:\Document\errorCSV\wrong_WriteTemp.CSV', delim = ',', header = True, parallel = true, null_padding = true);
┌───────┬───────┬─────────┐
│  c1   │  c2   │ column2 │
│ int64 │ int64 │ varchar │
├───────┼───────┼─────────┤
│     1 │     2 │         │
│     3 │     4 │         │
└───────┴───────┴─────────┘
D Select * From read_csv('D:\Document\errorCSV\wrong_WriteTemp.CSV', delim = ',', header = True, parallel = false, null_padding = true);
┌───────┬───────┬─────────┐
│  c1   │  c2   │ column2 │
│ int64 │ int64 │ varchar │
├───────┼───────┼─────────┤
│     1 │     2 │         │
│     3 │     4 │         │
│       │       │         │
└───────┴───────┴─────────┘
```

**Reply by szarnyasg** [reply]
Thanks, that extra line indeed shouldn't be there. I'll file a bug report.
